File: app/src/main/java/com/drdisagree/iconify/utils/ObservableVariable.java
package com.drdisagree.iconify.utils;

public class ObservableVariable<T> {

    private T value;
    private OnChangeListener<T> listener;

    public void setOnChangeListener(OnChangeListener<T> listener) {
        this.listener = listener;
    }

    public void notifyChanged() {
        if (listener != null) {
            listener.onChange(value);
        }
    }

    public T getValue() {
        return value;
    }

    public void setValue(T newValue) {
        if (value != newValue) {
            value = newValue;
            if (listener != null) {
                listener.onChange(value);
            }
        }
    }

    public interface OnChangeListener<T> {
        void onChange(T newValue);
    }
}

File: app/src/main/java/com/drdisagree/iconify/utils/extension/ObservableVariable.kt
package com.drdisagree.iconify.utils.extension

class ObservableVariable<T> {

    var value: T? = null
        private set
    private var listener: OnChangeListener<T?>? = null

    fun setOnChangeListener(listener: OnChangeListener<T?>?) {
        this.listener = listener
    }

    fun notifyChanged() {
        listener?.onChange(value)
    }

    fun setValue(newValue: T) {
        if (value !== newValue) {
            value = newValue

            listener?.onChange(value)
        }
    }

    fun interface OnChangeListener<T> {
        fun onChange(newValue: T)
    }
}
